waltz-client/src/main/java/com/wepay/waltz/client/internal/Partition.java

```diff
@@ -236,24 +239,39 @@ public void unmounted(WaltzNetworkClient networkClient) {
             if (this.networkClient == networkClient) {
                 logger.debug("partition unmounted: {}", this);
                 this.mounted = false;
+                this.clientHWMAhead = false;
                 this.networkClient = null;
                 lock.notifyAll();
             }
         }
     }
 
+    /**
+     * Invoked after receiving MOUNT_RESPONSE informing that local partition' HWM is ahead of server's HWM.
+     * If thread is stuck in ensureMounted, this thread is woken up and throws IllegalStateException
+     */
+    public void partitionAhead() {
+        synchronized (lock) {
+            this.clientHWMAhead = true;
```

Comment:
Let's assume that the client's HWM is ahead of server and partitionAhead() is triggered. This will set the clientHWMAhead as True. And if the client the thread is stuck in "ensureMounted()", it will throw an exception.

How are making sure that eventually (after updating the HWM table) the client will  be able to mount this partition successfully ?

Lets assume there is a transaction written to this partition so "WaltzClient.submit(..)" will be called and eventually "ensureMounted()" will be triggered. But the ensureMounted() will throw an exception again as "clientHWMAhead is set to True". And also note that the client didn't send any MOUNT_REQUEST message again to server in order to be able to mount the partition successfully.